datafusion/optimizer/src/decorrelate_scalar_subquery.rs

```diff
@@ -0,0 +1,360 @@
+            relation: Some(subqry_alias.clone()),
+            name: it.name.clone(),
+        })
+        .collect();
+    let join_keys = (outer_cols, subqry_cols);
+
+    // join our sub query into the main plan
+    let new_plan = LogicalPlanBuilder::from(filter_input.clone());
+    let new_plan = if join_keys.0.is_empty() {
+        // if not correlated, group down to 1 row and cross join on that (preserving row count)
+        new_plan.cross_join(&subqry_plan)?
+    } else {
+        // inner join if correlated, grouping by the join keys so we don't change row count
+        new_plan.join(&subqry_plan, JoinType::Inner, join_keys, None)?
+    };
+
+    // if the main query had additional expressions, restore them
+    let new_plan = if let Some(expr) = combine_filters(outer_others) {
+        new_plan.filter(expr)?
+    } else {
+        new_plan
+    };
+
+    // restore conditions
+    let qry_expr = Box::new(Expr::Column(Column {
+        relation: Some(subqry_alias),
+        name: "__value".to_string(),
+    }));
+    let filter_expr = if query_info.expr_on_left {
+        Expr::BinaryExpr {
+            left: Box::new(query_info.expr.clone()),
+            op: query_info.op,
+            right: qry_expr,
+        }
+    } else {
+        Expr::BinaryExpr {
+            left: qry_expr,
+            op: query_info.op,
+            right: Box::new(query_info.expr.clone()),
+        }
+    };
+    let new_plan = new_plan.filter(filter_expr)?;
+
+    let new_plan = new_plan.build()?;
+    Ok(Some(new_plan))
+}
+
+struct SubqueryInfo {
+    query: Subquery,
+    expr: Expr,
+    op: Operator,
+    expr_on_left: bool,
+}
+
+impl SubqueryInfo {
+    pub fn new(query: Subquery, expr: Expr, op: Operator, expr_on_left: bool) -> Self {
+        Self {
+            query,
+            expr,
+            op,
+            expr_on_left,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test::*;
+    use datafusion_expr::{col, logical_plan::LogicalPlanBuilder, min, Operator, scalar_subquery};
+    use datafusion_common::{Column, Result};
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let rule = DecorrelateScalarSubquery::new();
+        let optimized_plan = rule
+            .optimize(plan, &mut OptimizerConfig::new())
+            .expect("failed to optimize plan");
+        let formatted_plan = format!("{}", optimized_plan.display_indent_schema());
+        assert_eq!(formatted_plan, expected);
+    }
+
+    /// Test for correlated scalar subquery filter
+    #[test]
+    fn exists_subquery_correlated() -> Result<()> {
+        let sq = Arc::new(
+            LogicalPlanBuilder::from(test_table_scan_with_name("sq")?)
+                .filter(Expr::BinaryExpr {
+                    left: Box::new(Expr::Column(Column::from("test.a"))),
+                    op: Operator::Eq,
+                    right: Box::new(Expr::Column(Column::from("sq.a"))),
+                })?
```

Comment:
This can be written more concisely:

```suggestion
                .filter(col("test.a").eq(col("sq.a")))?
```